### PR TITLE
PSMove: initial support emulate psmove with mouse handler.

### PR DESCRIPTION
**Pull request by Zangetsu38**
- Credit to @VelocityRa for Original commit with initial mouse emulate psmove support here VelocityRa@55c5b2a

- Initial implement support mouse with use mouse handler for emulate cursor PSMove, good works particularly for rail shooter game like house of the dead serie and time crisis.

- Fix some warning with value uninitialized.
- Disable double click full screen if enable mouse psmove handler.

After used my long test with implemented real PSMove tracker for understand how use this for implemented function pos.

# For tester, button on mouse is:
- Left click = Trigger
- Right click = Move
- Middle click = X
- Middle + Left = Start
- Middle + right = Triangle
- Left + right = Cirle

# Configuration request
image

### CODE REVIEW

**Comment by vsub** [comment]
Will this eventually allow Ar Tonelico Quga "shaking the gamepad" to work with the mouse?
I tried it but nothing happens(if I set everything right)

**Comment by legend800** [comment]
Works great here (wish there was a way to shake):
2019-09-21 09_16_38-FPS_ 41 51 _ Vulkan _ 0 0 7-87f66a8d-psmove-mouse _ THE HOUSE OF THE DEAD 4  NPU

How to navigate - is there ANY other input besides the 3 mouse buttons?


HOD Overkill is black screen - fine on master:
2019-09-21 09_21_36-Window

Can't test - broken on master too. I'm tracking that down now. 
2019-09-21 09_19_59-Window

**Comment by Zangetsu38** [comment]
For HOD test it with fake psmove in master for compare

**Comment by legend800** [comment]
Confirmed it only hangs with Move Handler = Mouse. With Fake, it loads fine, like master, so no regression. Updated post. 

Sucks that Time Crisis was broken 5 days ago since I think that's the one rail shooter that will work perfectly with the current ps move implementation.

**Comment by Zangetsu38** [comment]
for me  i get here
image
but calibrate not works

**Comment by legend800** [comment]
Strange. Only boots if I set move handler to null. Maybe it's a US vs non-US version of game.

**Comment by toccata10** [comment]
I was hoping for this to revive some railshooters. Thanks for your work.
It allowed me to see a bit of Maddog for the 1st time: the haircross moves, I can shoot. If I'm in Fullscreen, I can reload by firing outside the screen.
Unfortunately, after some time, either:
*the game freezes (it's still on, but you can't go to the next sequence)
*the mouse stops controlling the haircross (and the firing)
Capture_067
A few other tests: 
*Hotd4: can't reload, so the game ends quite quickly.
*Hotd overkill: by middle clicking, I can get past the calibration and launch a mini-game. In the mini-game, the mouse doesn't respond.
Capture_069
Capture_071

**Comment by Zangetsu38** [comment]
Re test needed

**Comment by RiasatSalminSami** [comment]
Yeah, it works very well. There needs to be some functions for shaking too. Some games reload by shaking like house of dead 4 and on hod4, some parts of the levels can't even be progressed without shaking.

So there has to be a way to assign keyboard and mouse buttons for Psmove.

Lmb= Trigger= Shoot
Rmb= Shake = Reload/   Aim outside and shoot= Reload

And so on.

Without the ability to emualte Psmove shaking and assign the buttons, it's gonna be very limited. However, very good progress.

**Comment by Zangetsu38** [comment]
for full screen is normal, you need re calibrate cursor for that, but i working on one things for can keep same calibrate any resolution using

**Comment by legend800** [comment]
Now that Time Crisis is fixed, I gave it a go with this PR. Unfortunately, the US version is unplayable as it needs Triangle mapping for calibration and it won't let you play unless calibrated. 

Is there anyway to map Triangle to something else or allow keyboard mapping in addition? Maybe a setting in the config file even? Could you post a build maybe with Triangle instead of X button for further testing in the meantime?

2019-09-24 08_20_34-FPS_ 60 03 _ Vulkan _ 0 0 7-e0b90b7c-psmove-mouse _ TIME CRISIS_ RAZING STORM  B

2019-09-24 08_23_47-Pull Requests · RPCS3_rpcs3

**Comment by Zangetsu38** [comment]
@legend800 pres Triangle on ds3 input not works ?

**Comment by legend800** [comment]
Tried, unfortunately input is not recognized by other devices here, only the device being calbrated. :(

**Comment by Zangetsu38** [comment]
@legend800 i see, after confirm calibrate i can use circle for back on menu, but is diferent so hmm

**Comment by legend800** [comment]
Yep, input is locked to single device in the cali screen. BTW, this is Wes from discord. ;)

**Comment by Zangetsu38** [comment]
re test needed, you can now change size game window and no need recalibrate or other things

**Comment by Zangetsu38** [comment]
@legend800 retry please for your USA version of time crisis, and use press Midle click + right click for triangle.

**Comment by legend800** [comment]
Nice! Works great BUT I can't escape cali screen without O button as well unfortunately. I can confirm cali screen now, but then it just takes me back to cali other pads instead of proceeding to main menu. See below screens.

SO, can you do that trick on the Middle click + left click for O button? Then we would have X, O and Triangle all accessible. :)

2019-09-26 20_12_38-FPS_ 59 88 _ Vulkan _ 0 0 7-14549b3e-psmove-mouse _ TIME CRISIS_ RAZING STORM  B

2019-09-26 20_10_10-FPS_ 34 11 _ Vulkan _ 0 0 7-14549b3e-psmove-mouse _ TIME CRISIS_ RAZING STORM  B

**Comment by legend800** [comment]
Scratch the above request - you can use other controllers to exit the screen thankfully - in my case I use Pad1 for O button to exit. 

Also, you already have Middle click + left click mapped - Start button! Which actually is required for all the titles in Razing storm to skip cinematics, pause and most importantly, continue after dying. 

So don't touch a thing, everything seems perfect now, and all required button functionality is accessible. Nice work! Hopefully you can get it reviewed and merged soon.

**Comment by Zangetsu38** [comment]
Is now Ready for review

**Comment by Zangetsu38** [comment]
> 
> 
> Scratch the above request - you can use other controllers to exit the screen thankfully - in my case I use Pad1 for O button to exit.
> 
> Also, you already have Middle click + left click mapped - Start button! Which actually is required for all the titles in Razing storm to skip cinematics, pause and most importantly, continue after dying.
> 
> So don't touch a thing, everything seems perfect now, and all required button functionality is accessible. Nice work! Hopefully you can get it reviewed and merged soon.

Yes, i know, i have explain this just before, after confirm calibrate you can use ds3 input for back

**Comment by dudesh34** [comment]
Hi,
 what i need to put in pad and in i/o settings for this to work? 
thanks.

**Comment by Zangetsu38** [comment]
> 
> 
> Hi,
> what i need to put in pad and in i/o settings for this to work?
> thanks.
> i will try it with a dolphin bar and will update.

eu what ? juust set Mouse on I/o tab inside PSMove handler, and that all, and dolphin bar ?

**Comment by legend800** [comment]
It actually needs 3 settings changed to work:
2019-09-27 18_50_27-Settings

Zangetsu, you should consider updating those tooltips on that window at some point to show that it IS now implemented (says ps move is not supported currently).

**Comment by Zangetsu38** [comment]
@legend800 error, you need set Camera setting on fake

File: rpcs3/Emu/Cell/Modules/cellGem.cpp [hunk 1]
```diff
@@ -299,6 +285,112 @@ static bool map_ext_to_ds3_input(const u32 port_no, CellGemExtPortData& ext)
...
+	const auto renderer = static_cast<GSRender*>(rsx::get_current_renderer());
+	const auto width = renderer->get_frame()->client_width();
+	const auto hight = renderer->get_frame()->client_height();
+
+	const f32 scaling_width = static_cast<f32>(width) / 640;
+	const f32 scaling_hight = static_cast<f32>(hight) / 480;
+	const f32 x = static_cast<f32>(mouse.x_pos) / scaling_width;
+	const f32 y = static_cast<f32>(mouse.y_pos) / scaling_hight;
+
+	gem_state->quat[0] = -x + 320;
```
Comment (elad335) [inline rpcs3/Emu/Cell/Modules/cellGem.cpp]:
```suggestion
	gem_state->quat[0] = 320.f - x;
```

File: rpcs3/Emu/Cell/Modules/cellGem.cpp [hunk 2]
```diff
@@ -299,6 +285,112 @@ static bool map_ext_to_ds3_input(const u32 port_no, CellGemExtPortData& ext)
...
+	gem_state->quat[1] = static_cast<f32>(mouse.y_pos) / scaling_width - 180;
```
Comment (elad335) [inline rpcs3/Emu/Cell/Modules/cellGem.cpp]:
```suggestion
	gem_state->quat[1] = (mouse.y_pos / scaling_width) - 180;
```

File: rpcs3/Emu/Cell/Modules/cellGem.cpp [hunk 3]
```diff
@@ -299,6 +285,112 @@ static bool map_ext_to_ds3_input(const u32 port_no, CellGemExtPortData& ext)
...
+		return false;
+	}
+
+	const auto& mouse = handler->GetMice().at(0);
+
+	const auto renderer = static_cast<GSRender*>(rsx::get_current_renderer());
+	const auto width = renderer->get_frame()->client_width();
+	const auto hight = renderer->get_frame()->client_height();
+
+	const f32 scaling_width = static_cast<f32>(width) / 640;
```
Comment (elad335) [inline rpcs3/Emu/Cell/Modules/cellGem.cpp]:
```suggestion
	const f32 scaling_width = width / 640.f;
```

Comment (elad335) [reply rpcs3/Emu/Cell/Modules/cellGem.cpp]: Same for the line below.

File: rpcs3/Emu/Cell/Modules/cellGem.cpp [hunk 4]
```diff
@@ -299,6 +285,112 @@ static bool map_ext_to_ds3_input(const u32 port_no, CellGemExtPortData& ext)
...
+	const auto handler = g_fxo->get<MouseHandlerBase>();
+
+	std::lock_guard lock(handler->mutex);
+
+	if (!handler || mouse_no >= handler->GetMice().size())
+	{
+		return false;
+	}
+
+	memset(&digital_buttons, 0, sizeof(digital_buttons));
```
Comment (elad335) [inline rpcs3/Emu/Cell/Modules/cellGem.cpp]: This needs a temp var.
```suggestion
	const auto& mouse_data = handler->GetMice().at(0);

	u16 buttons_data = 0;
	if (mouse_data.buttons & CELL_MOUSE_BUTTON_1)
		buttons_data |= CELL_GEM_CTRL_T;
	if (mouse_data.buttons & CELL_MOUSE_BUTTON_2)
		buttons_data |= CELL_GEM_CTRL_MOVE;
	if (mouse_data.buttons & CELL_MOUSE_BUTTON_3)
		buttons_data |= CELL_GEM_CTRL_CROSS;
	if (mouse_data.buttons & CELL_MOUSE_BUTTON_3 && mouse_data.buttons & CELL_MOUSE_BUTTON_2)
		buttons_data |= CELL_GEM_CTRL_TRIANGLE;
	if (mouse_data.buttons & CELL_MOUSE_BUTTON_3 && mouse_data.buttons & CELL_MOUSE_BUTTON_1)
		buttons_data |= CELL_GEM_CTRL_START;
	digital_buttoms = buttons_data;
```

**Comment by toccata10** [comment]
This morning, I made some tests with all the railshooters I know on ps3.
Results:

*Deadstorm pirates (NPEB00409):
Combined with a pad configured, I'm able to enter the menus (for example, I press the Circle on my controller to avoid connecting to Playstation Network), do the calibration and enter the mode selection (pressing circle).
The mouse is recognized as "Calibrate Motion Controller n°7"
When in full screen, I can't see the mouse pointer, so I can't calibrate.
When the window is maximized, it's working well. Whell I'm in fullscreen, I must make big moves with the mouse to move the crosshair. I tried to increase "mouse acceleration" to 20 and deadzone to 0, but it didn't change anything. The crosshair moves slowly when in fullscreen. I'm on a 4k screen (don't know if that's relevant).
The game plays nicely (remember to active "write color buffers").

*Mad dog mccree: calibration ok, no slowdown in fullscreen. I'm able to reload the gun in fullscreen by firing outside the screen. Pb: Game freezes while in the main street. 
*Maddog 2 & The last bounty hunter: same behaviour.
*Fast Draw showdown: can't calibrate (the crosshair stays on the left side). Strange as it seems the same engine as MadDog.

*Heavyfire: afghanistan: stuck on "select profile-player 2"
*Heavyfire: shattered spea: stuck on "select profile-player 2"

*Men In Black: alien crisis: can't navigate the menus with the pad. I'm able to launch a game, but it asks to press L1 which I can't.

*House of the Dead 3: very similar to Deadstorm pirates: calibration ok, but cursor very slow in fullscreen. Reload is automatic. With Vulkan: poor performance and exploding vertices. With Ogl: ok. It played nicely with window maximized until the mouse stopped controlling the crosshair (the game was still playing, but I had no control)

*House of the Dead 4: looks like hotd3: performance is much better in Ogl. Pb is you can't reload (you need to shake the gun for that).

*House of the dead overkill: for the calibration, I can fire with the Middle click (X button), not the left one. The control works until I launch a minigame: it stops responding then. tried with and without WCB: no change.

*the shoot: can't calibrate

*time crisis: razing storm: can use the mouse and a pad to navigate the options. Crosshair speed is good in windowed and fullscreen. I can play the game (reload with the right button). It played nicely until the mouse stopped controlling the crosshair (the game was still playing, but I had no control).

*time crisis 4 BLUS30093: mouse not recognized.

**Comment by Zangetsu38** [comment]
thx for your test, i know is not pertfect
- Men in blac game is not compatible with psmove so don't relate this pr.
- DeadStorm pirate and HOD 3 perfectly works for me.
- HOD request special things, with real psmove is same, shack controller no works, (works on HOD3) so using different function not implemented for that.
- For HOD overkill i have not found where is probleme, i need test with real psmove for compare
- The shoot have very old graphic issue with cursor missing, same for real psmove.
- For razing storm i don't have see any probleme, full works.
- Time Crsis 4 works, just i have 60 fps but game run in slowdown, no idea why.
- Heavyfire: afghanistan is stupid, it ask connect all user if i using psmove in port 7 
- for both resident evil rail shooter, calibrate works, but cursor no move, i need look whaty function used for it
- Resident evil 5 and resistance 3 full works

**Comment by Andryshik345** [comment]
Sports Champions 2 asking for PS Eye connecting, Sports Champions is softlocks just after logo.
изображение
[RPCS3.log.gz](https://github.com/RPCS3/rpcs3/files/3676385/RPCS3.log.gz)

**Comment by Zangetsu38** [comment]
for sport champion 1/2 ignore it, it not can works with mouse, need obligatory real psmove

**Comment by Zangetsu38** [comment]
good news, founded what function is used for resident evil, exactly like i have thinks
image

**Comment by Zangetsu38** [comment]
@toccata10 pushed, with news change, you can news try both resident evil, works for me perfect after undertund how good calibrate

**Comment by toccata10** [comment]
I wonder what is the trick to calibrate in resident evil ? How do press on Circle to exit the calibration screen ?

**Comment by Zangetsu38** [comment]
> I wonder what is the trick to calibrate in resident evil ? How do press on Circle to exit the calibration screen ?

@toccata10 look description for circle, and for calibrate is special, no press triger in center screen, but
in little decaled up right in end circle 
like that, if you calibreate normally, you set psmove cursor in center, and repress T for re calibrate
result

- First calibrate:  
image

- Second calibrate
image

**Comment by Zangetsu38** [comment]
for second game is very easy, because first calibreate already show target, so just click here for first click
image

**Comment by Zangetsu38** [comment]
@toccata10 you can test agaion please, on resident evil serie ?

**Comment by toccata10** [comment]
I was stuck because I didn't see that left+right sent "circle".
So, ok, it's working. Nice.
About the cursor being very slow in fullscreen, here's a workaround:
with the mouse, click on an area near the center instead of click on the target (see the red spot).
Capture_076b

I can't remember exactly but I think that I can't throw grenade because I have to press 2 buttons (left+right ?) but as it's associated with circle, it doesn't work.
Nice work you've done.
An idea (if that's possible): allow the keyboard to send some presses (instead of combinations middle+left,...). 

No luck with time crisis (with the razing storm edition, as it's the only one supporting psmove): I can't move the haircross in the game (but I can calibrate).

**Comment by Zangetsu38** [comment]
@toccata10  Yes, i know i thinks for found solution for this 🤔

**Comment by dudesh34** [comment]
do i need to asign mouse movements and buttons  in pad or just the after make
 the settings in i/o? or not to touch anything in pad?

**Comment by legend800** [comment]
Just settings in I/O.

In additional, you can use the Pad, mapped to keyboard, to do generic actions (like main menu nav).

**Comment by Zangetsu38** [comment]
for pad is bad idea, is like fake actually that coded problem and conflict with ds3 pad, i can only use so mouse + keyboard

**Comment by toccata10** [comment]
A bit more testing:
Razing storm [blus30528] contains 3 games
*Razing Storm: time crisis
Need a pad configured
Navigate menus (arrows) with keyboard or pad, validate with mouse 
Once calibration is done, it's a bit difficult (for me) to press middle+right to send Triangle, so I first press right, then middle while maintaining right.
After that, I press circle on the keyboard.
I launch the game. To bypass the cinematic: left+middle.
After that, it's ok: left to fire. Right to reload.
Ok but freezes after some time (the mouse no longer controls the haircross)

Time crisis:
Same calibration. Choose "Motion controller 7".
Works fine but freezes after some time (the mouse no longer controls the haircross)

Deadstorm pirates
Same calibration
Plays nice. Didn't freeze.

**Comment by dudesh34** [comment]
thanks for all the updates, cant wait for mor improvments:)

**Comment by dudesh34** [comment]
Hi, is there a new update to download? new fixes?
thanks.

**Comment by toccata10** [comment]
I've tried playing Deadstorm on Linux with a wiimote: it's rather cool. Thing is: it randomly stops considerig the mouse after a while: the game is still running, but the cursor doesn't answer to the mouse anymore.
Same behaviour in Razing storm.
I uploaded my (big) logfile, just in case someone can find a clue in it:
https://www.dropbox.com/s/426bajbwgusj76j/RPCS3.log-deadstorm.gz?dl=0

**Comment by Zangetsu38** [comment]
@toccata10 i have not see this issue for moment, play How long before get this ?

**Comment by toccata10** [comment]
It depends: I played 1 mission in Deadstorm, then I started another one and it froze a few seconds after that.
With Razing storm, it froze quite quickly (after 2-3 minutes in the play).
 It seems random.

**Comment by toccata10** [comment]
Another test this morning running through gdb: I launched Razing Storm. After calibration, I started the game: it crashed on "loading". Here's the backtrace:
(gdb) bt
#0  0x00000000424de484 in __0x6c842c ()
#1  0x000000000070c7ff in ppu_thread::exec_task() ()
#2  0x000000000070cf56 in ppu_thread::fast_call(unsigned int, unsigned int) ()
#3  0x0000000000712b0f in ppu_thread::cpu_task() ()
#4  0x00000000009c117f in cpu_thread::operator()() ()
#5  0x00000000007068f6 in named_thread<ppu_thread>::entry_point(void*) ()
#6  0x00007ffff78cf4cf in start_thread () from /usr/lib/libpthread.so.0
#7  0x00007ffff57f22d3 in clone () from /usr/lib/libc.so.6
(gdb)

**Comment by elad335** [comment]
Need to solve conflcrs.

**Comment by toccata10** [comment]
Just tested Razing Storm again: same behaviour: Ok but freezes after some time (the mouse no longer controls the haircross).

**Comment by legend800** [comment]
Never experienced that issue. Hard to say if 100% related to this implementation. I tried the latest build and like with every previous build, input never cut out, even after playing half the game (beat 2/ 4 stages). It was perfect and I hope we can just merge this finally and continue to iterate on it. 



2019-12-01 18_08_59-FPS_ 59 95 _ Vulkan _ 0 0 7-3c48881a-psmove-mouse _ TIME CRISIS_ RAZING STORM  B

**Comment by AlphaDelta9** [comment]
Tested BLUS30528TC4 freezes after long time ,mouse no longer control the crosshair and not responding. Already lost 2x my progress at Stage 2 :(

File: rpcs3/Emu/Cell/Modules/cellGem.h [hunk 5]
```diff
@@ -139,16 +139,16 @@ struct CellGemExtPortData
...
+	be_t<f32> u;               // horizontal screen position in pixels
+	be_t<f32> v;               // vertical screen position in pixels
+	be_t<f32> r;               // size of sphere on screen in pixels
 	be_t<f32> projectionx;
 	be_t<f32> projectiony;
-	be_t<f32> distance;
-	u8 visible;
-	u8 r_valid;
+	be_t<f32> distance;        // Move sphere distance from camera (probably)
+	u8 visible;	               // whether the sphere is visible in the current frame
```
Comment (Megamouse) [inline rpcs3/Emu/Cell/Modules/cellGem.h]: What's up with the whitespace here?

File: rpcs3/Emu/Cell/Modules/cellGem.cpp [hunk 6]
```diff
@@ -775,14 +882,21 @@ error_code cellGemGetState(u32 gem_num, u32 flag, u64 time_parameter, vm::ptr<Ce
...
+		ds3_input_to_pad(gem_num, gem_state->pad.digitalbuttons, gem_state->pad.analog_T);
+	else if (g_cfg.io.move == move_handler::mouse)
 	{
-		map_to_ds3_input(gem_num, gem_state->pad.digitalbuttons, gem_state->pad.analog_T);
-		map_ext_to_ds3_input(gem_num, gem_state->ext);
+		mouse_input_to_pad(gem_num, gem_state->pad.digitalbuttons, gem_state->pad.analog_T);
+		mouse_pos_to_gem_state(gem_num, gem_state);
+	}
+
+	if (!(g_cfg.io.move == move_handler::null))
```
Comment (Megamouse) [inline rpcs3/Emu/Cell/Modules/cellGem.cpp]: Remove this check and always execute the code below.
Move the nulll handler check above the other move handler checks and return CELL_GEM_NOT_CONNECTED if move_handler::null

File: rpcs3/Emu/Cell/Modules/cellGem.cpp [hunk 7]
```diff
@@ -645,14 +751,15 @@ error_code cellGemGetInertialState(u32 gem_num, u32 state_flag, u64 timestamp, v
...
 	if (g_cfg.io.move == move_handler::fake)
-	{
-		map_to_ds3_input(gem_num, inertial_state->pad.digitalbuttons, inertial_state->pad.analog_T);
-		map_ext_to_ds3_input(gem_num, inertial_state->ext);
+		ds3_input_to_pad(gem_num, inertial_state->pad.digitalbuttons, inertial_state->pad.analog_T);
+	else if (g_cfg.io.move == move_handler::mouse)
+		mouse_input_to_pad(gem_num, inertial_state->pad.digitalbuttons, inertial_state->pad.analog_T);
 
-		inertial_state->timestamp = gem->timer.GetElapsedTimeInMicroSec();
+	if (!(g_cfg.io.move == move_handler::null)) {
```
Comment (Megamouse) [inline rpcs3/Emu/Cell/Modules/cellGem.cpp]: Remove this check and always execute the code below.
Move the nulll handler check above the other move handler checks and return CELL_OK if move_handler::null

File: rpcs3/Emu/Cell/Modules/cellGem.cpp [hunk 8]
```diff
@@ -601,26 +702,31 @@ error_code cellGemGetImageState(u32 gem_num, vm::ptr<CellGemImageState> image_st
...
-		image_state->visible = true;
-		image_state->u = 0;
-		image_state->v = 0;
-		image_state->r = 20;
-		image_state->r_valid = true;
-		image_state->distance = 2 * 1000;   // 2 meters away from camera
-		// TODO
-		image_state->projectionx = 1;
-		image_state->projectiony = 1;
+	if (!(g_cfg.io.move == move_handler::null))
```
Comment (Megamouse) [inline rpcs3/Emu/Cell/Modules/cellGem.cpp]: Remove this check and always execute the code below.
Move the nulll handler check above the other move handler checks and return CELL_OK if move_handler::null

File: rpcs3/Emu/Cell/Modules/cellGem.cpp [hunk 9]
```diff
@@ -304,6 +290,119 @@ static bool map_ext_to_ds3_input(const u32 port_no, CellGemExtPortData& ext)
...
+
+	analog_t = mouse_data.buttons & CELL_MOUSE_BUTTON_1 ? 0xFFFF : 0;
+
+	return true;
+}
+
+static bool mouse_pos_to_gem_image_state(const u32 mouse_no, vm::ptr<CellGemImageState>& gem_image_state)
+{
+		// Only game this seems to work on is PAIN, others use different functions
+
```
Comment (Megamouse) [inline rpcs3/Emu/Cell/Modules/cellGem.cpp]: This whole function has wrong spacing

File: rpcs3/Emu/Cell/Modules/cellGem.cpp [hunk 10]
```diff
@@ -304,6 +290,119 @@ static bool map_ext_to_ds3_input(const u32 port_no, CellGemExtPortData& ext)
...
+	if (!handler || mouse_no >= handler->GetMice().size())
+	{
+		return false;
+	}
+
+	memset(&digital_buttons, 0, sizeof(digital_buttons));
+
+	const auto& mouse_data = handler->GetMice().at(0);
+
+	if (mouse_data.buttons & CELL_MOUSE_BUTTON_1 && mouse_data.buttons & CELL_MOUSE_BUTTON_2)
```
Comment (Megamouse) [inline rpcs3/Emu/Cell/Modules/cellGem.cpp]: Use () around the individual statements

File: rpcs3/Emu/Cell/Modules/cellGem.cpp [hunk 11]
```diff
@@ -304,6 +290,119 @@ static bool map_ext_to_ds3_input(const u32 port_no, CellGemExtPortData& ext)
...
+		digital_buttons |= CELL_GEM_CTRL_CIRCLE;
+	if (mouse_data.buttons & CELL_MOUSE_BUTTON_3)
+		digital_buttons |= CELL_GEM_CTRL_CROSS;
+	if (mouse_data.buttons & CELL_MOUSE_BUTTON_2)
+		digital_buttons |= CELL_GEM_CTRL_MOVE;
+	if (mouse_data.buttons & CELL_MOUSE_BUTTON_1 && mouse_data.buttons & CELL_MOUSE_BUTTON_3)
```
Comment (Megamouse) [inline rpcs3/Emu/Cell/Modules/cellGem.cpp]: Use () around the individual statements

File: rpcs3/Emu/Cell/Modules/cellGem.cpp [hunk 12]
```diff
@@ -304,6 +290,119 @@ static bool map_ext_to_ds3_input(const u32 port_no, CellGemExtPortData& ext)
...
+		digital_buttons |= CELL_GEM_CTRL_START;
+	if (mouse_data.buttons & CELL_MOUSE_BUTTON_1)
+		digital_buttons |= CELL_GEM_CTRL_T;
+	if (mouse_data.buttons & CELL_MOUSE_BUTTON_2 && mouse_data.buttons & CELL_MOUSE_BUTTON_3)
```
Comment (Megamouse) [inline rpcs3/Emu/Cell/Modules/cellGem.cpp]: Use () around the individual statements

File: rpcs3/Emu/Cell/Modules/cellGem.cpp [hunk 13]
```diff
@@ -304,6 +290,119 @@ static bool map_ext_to_ds3_input(const u32 port_no, CellGemExtPortData& ext)
...
+		digital_buttons |= CELL_GEM_CTRL_TRIANGLE;
+
+	analog_t = mouse_data.buttons & CELL_MOUSE_BUTTON_1 ? 0xFFFF : 0;
```
Comment (Megamouse) [inline rpcs3/Emu/Cell/Modules/cellGem.cpp]: Use () around the button check

File: rpcs3/Emu/Cell/Modules/cellGem.cpp [hunk 14]
```diff
@@ -304,6 +290,119 @@ static bool map_ext_to_ds3_input(const u32 port_no, CellGemExtPortData& ext)
...
+			return false;
+		}
+
+		const auto& mouse = handler->GetMice().at(0);
+
+		const auto renderer = static_cast<GSRender*>(rsx::get_current_renderer());
+		const auto width = renderer->get_frame()->client_width();
+		const auto hight = renderer->get_frame()->client_height();
+		const f32 scaling_width = width / 640.f;
+		const f32 scaling_hight = hight / 480.f;
```
Comment (Megamouse) [inline rpcs3/Emu/Cell/Modules/cellGem.cpp]: What are all these magic numbers in all the helper functions?

**Comment by legend800** [comment]
@AlphaDelta9 And that absolutely does not do that on master and this is 100% because of the PR? I don't want to block this PR because of false positives.

File: rpcs3/Emu/Cell/Modules/cellGem.cpp (hunk 14 above)
Comment (Zangetsu38) [reply rpcs3/Emu/Cell/Modules/cellGem.cpp]: is obligatory for scaling

File: rpcs3/Emu/Cell/Modules/cellGem.cpp (hunk 8 above)
Comment (Zangetsu38) [reply rpcs3/Emu/Cell/Modules/cellGem.cpp]: in master is already like that, run this only with fake, i have just set same

File: rpcs3/Emu/Cell/Modules/cellGem.cpp (hunk 7 above)
Comment (Zangetsu38) [reply rpcs3/Emu/Cell/Modules/cellGem.cpp]: same, on master is same only for fake, so here i just set fake/mouse get same value

File: rpcs3/Emu/Cell/Modules/cellGem.h (hunk 5 above)
Comment (Zangetsu38) [reply rpcs3/Emu/Cell/Modules/cellGem.h]: no idea, file is good for me for have all line aligned

File: rpcs3/Emu/Cell/Modules/cellGem.cpp (hunk 6 above)
Comment (Zangetsu38) [reply rpcs3/Emu/Cell/Modules/cellGem.cpp]: move handler null already get gem not connected, i have initialized it in first
here is same, master already run code if have fake, i get just same

**Comment by AlphaDelta9** [comment]
> 
> 
> @AlphaDelta9 And that absolutely does not do that on master and this is 100% because of the PR? I don't want to block this PR because of false positives.

@legend800 my bad not compare it with master build, i just play with this build because i want to give some feedback. I want too this build merge with master, don't mean to give false postive. I'm sorry

**Comment by dudesh34** [comment]
hi, is there a new update to download in last month? any news on merge?

File: rpcs3/Emu/Cell/Modules/cellGem.cpp (hunk 13 above)
Comment (Zangetsu38) [reply rpcs3/Emu/Cell/Modules/cellGem.cpp]: ah, i have just set same master, is velo code, not touched this

File: rpcs3/Emu/Cell/Modules/cellGem.cpp (hunk 12 above)
Comment (Zangetsu38) [reply rpcs3/Emu/Cell/Modules/cellGem.cpp]: is important ? 🤔

File: rpcs3/Emu/Cell/Modules/cellGem.h (hunk 5 above)
Comment (elad335) [reply rpcs3/Emu/Cell/Modules/cellGem.h]: NEVER use tabs for stuff like this, this is the result.

Comment (Zangetsu38) [reply rpcs3/Emu/Cell/Modules/cellGem.h]: ah ok

Comment (Zangetsu38) [reply rpcs3/Emu/Cell/Modules/cellGem.h]: @elad335 after check, velocity have used space here, so where is problem ?

Comment (Zangetsu38) [reply rpcs3/Emu/Cell/Modules/cellGem.h]: image

Comment (Zangetsu38) [reply rpcs3/Emu/Cell/Modules/cellGem.h]: image

Comment (elad335) [reply rpcs3/Emu/Cell/Modules/cellGem.h]: odd, it's misaligned for me.

Comment (Zangetsu38) [reply rpcs3/Emu/Cell/Modules/cellGem.h]: juust ignore it on gihub, is perfect

**Comment by Zangetsu38** [comment]
all change request pushed

File: rpcs3/Emu/Cell/Modules/cellGem.h (hunk 5 above)
Comment (Megamouse) [reply rpcs3/Emu/Cell/Modules/cellGem.h]: image

Comment (Zangetsu38) [reply rpcs3/Emu/Cell/Modules/cellGem.h]: Ouch not see it lol to small

**Comment by rommel5888** [comment]
I am a Chinese Beginner
Left click = Trigger in time crisis is invalid！！！
Who is willing to help me！！！
thank！！！！

**Comment by ArcZac** [comment]
Somebody help!!

left clickk in Time Crisis 4 and Time Crisis Razing Storm doesn't work
i cant duck or shield , it just like playing without cover and shooting all the way like deadstorm pirate

In Deadstorm Pirate the Left Click working no problem at all, only in TC Razing Storm and TC 4 have a problem

**Comment by dudesh34** [comment]
Yes...left click doesnt work with time crisis 4 and razing storm..just tested it also.
any fix? thanks:)

**Comment by legend800** [comment]
yep broken, I'll log a bug for it

**Comment by dudesh34** [comment]
> yep broken, I'll log a bug for it

thanks man:)

**Comment by dudesh34** [comment]
So...i read someone fix it. did you try it?

**Comment by legend800** [comment]
now it's crashing, no idea what's going on in your config. Plays perfectly here.

**Comment by dudesh34** [comment]
ok, thanks...maybe my poor i5 6600 cant handle it.

**Comment by ArcZac** [comment]
> ok, thanks...maybe my poor i5 6600 cant handle it.

Can you play it at Vulkan GPU Settings? 720p?

**Comment by dudesh34** [comment]
i can play it without the ps move "hack"...its not perfect has i have a lot of sttuters and glitchy sounds.

**Comment by IlDucci** [comment]
The mouse cursor isn't detected when playing Dead Space Extraction with the PS Move hack. The buttons work, but the sight won't move.

**Comment by legend800** [comment]
@llDucci Can you add that comment in the new tracking issue here? https://github.com/RPCS3/rpcs3/issues/7840

**Comment by ElektroStudios** [comment]
someone tested Kung Fu Rider?.

**Comment by dudesh34** [comment]
Too bad it doesnt work with a light gun like aimtrak or wiimote ....the cursor just go all over the place:\
also, for me, time crisis 4 doesnt work is it anymore, anyone tried it lately?

**Comment by Megamouse** [comment]
stop chatting on old pull requests

**Pull request merged**